Sources/NIOHTTP1/HTTPObjectAggregator.swift

```diff
@@ -0,0 +1,408 @@
+    public typealias InboundOut = HTTPServerRequestFull
+
+    // Aggregator may generate responses of its own
+    public typealias OutboundOut = HTTPServerResponsePart
+
+    // TODO: could hold onto the ByteBuffers read from the channel instead of creating a single new one?
+    private var buffer: ByteBuffer! = nil
+    private var maxContentLength: Int
+    private var closeOnExpectationFailed: Bool
+    private var state: AggregatorState
+    
+    private var fullMessageHead: HTTPRequestHead! = nil
+    
+    public init(maxContentLength: Int, closeOnExpectationFailed: Bool = false) {
+        if maxContentLength < 0 {
+            preconditionFailure("maxContentLength must not be negative")
+        }
+        self.maxContentLength = maxContentLength
+        self.closeOnExpectationFailed = closeOnExpectationFailed
+        self.state = .idle
+    }
+
+    public func channelRead(context: ChannelHandlerContext, data: NIOAny) {
+        guard self.state != .closed else {
+            context.fireErrorCaught(HTTPObjectAggregatorError.connectionClosed)
+            return
+        }
+
+        let msg = self.unwrapInboundIn(data)
+        var serverResponse: HTTPResponseHead? = nil
+
+        switch msg {
+        case .head(let httpHead):
+            self.state.messageHeadReceived()
+            serverResponse = beginAggregation(context: context, request: httpHead, message: msg)
+        case .body(var content):
+            if self.state == .receiving {
+                serverResponse = aggregate(context: context, content: &content, message: msg)
+            }
+        case .end(let trailingHeaders):
+            endAggregation(context: context, trailingHeaders: trailingHeaders)
+            self.state.messageEndReceived()
+        }
+
+        // Generated a serverr esponse to send back
+        if let response = serverResponse {
+            context.write(self.wrapOutboundOut(.head(response)), promise: nil)
+            context.writeAndFlush(self.wrapOutboundOut(.end(nil)), promise: nil)
+        }
+
+        if self.state == .closed {
+            context.close(promise: nil)
+        }
+    }
+
+    func beginAggregation(context: ChannelHandlerContext, request: HTTPRequestHead, message: InboundIn) -> HTTPResponseHead? {
+        self.fullMessageHead = request
+        if let response = generateContinueResponse(context: context, request: request) {
+            if self.shouldCloseAfterContinueResponse(response: response) {
+                self.state.closed()
+            } else {
+                // Won't close the connection, but handling an oversized message
+                self.state.handlingOversizeMessage()
+                // Remove `Expect` header from the aggregated message
+                self.fullMessageHead?.headers.remove(name: "expect")
+            }
+            return response
+        } else if request.hasContentLength && request.contentLength! > self.maxContentLength {
```

Comment:
No need to check `hasContentLength` here either, just use the conditional binding.